### PR TITLE
Modify "Find" Function

### PR DESCRIPTION
**Pull request by JiaqingTan**
"Find" feature has been updated to include finding rooms and guests.

Examples are as below:
Guests: find -g n/Alex Yu p/86543322 e/abc@gmail.com t/frienderino
Rooms: find -r r/003 c/2 -hb from/ 01/02/2018 to/ 02/03/2018

Points to note:

- All filters after flags can be mixed and match.

- Room now had additional flags "-hb/-nb", has bookings and no bookings, as filter. To specify a date where room has bookings, simply add from and to range immediately after those flags. No specified date will mean that it shows rooms with all bookings / no bookings only.

### CODE REVIEW

File: src/main/java/seedu/address/logic/parser/FindCommandParser.java [hunk 1]
```diff
@@ -1,33 +1,312 @@
...
+        roomPredicates = new LinkedList<Predicate<Room>>();
+    }
+
     /**
      * Parses the given {@code String} of arguments in the context of the FindCommand
      * and returns an FindCommand object for execution.
+     *
      * @throws ParseException if the user input does not conform the expected format
      */
     public FindCommand parse(String args) throws ParseException {
```
Comment (adamwth) [inline src/main/java/seedu/address/logic/parser/FindCommandParser.java]: suggest using ArgumentTokenizer for this

**Comment by adamwth** [comment]
Coverage decreased by too much as well. Remember to write tests to cover for the new features you wrote!

File: src/main/java/seedu/address/logic/commands/AddCommand.java [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 import static java.util.Objects.requireNonNull;
 import static seedu.address.commons.util.CollectionUtil.requireAllNonNull;
+
```
Comment (teowz46) [inline src/main/java/seedu/address/logic/commands/AddCommand.java]: this didn't fail checkstyle?

File: src/main/java/seedu/address/model/guest/GuestTagsContainsKeywordsPredicate.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    private final List<String> keywords;
+
+    public GuestTagsContainsKeywordsPredicate(List<String> keywords) {
+        this.keywords = keywords;
+    }
+
+    @Override
+    public boolean test(Guest guest) {
+        List<Tag> guestTags = new LinkedList<>();
+        guestTags.addAll(guest.getTags());
```
Comment (teowz46) [inline src/main/java/seedu/address/model/guest/GuestTagsContainsKeywordsPredicate.java]: why not just `new LinkedList<>(guest.getTags())`

File: src/main/java/seedu/address/model/guest/GuestTagsContainsKeywordsPredicate.java [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+        for (String keywordTag : keywords) {
+            for (Tag guestTag : guestTags) {
+                if (guestTag.tagName.equals(keywordTag)) {
+                    return true;
```
Comment (teowz46) [inline src/main/java/seedu/address/model/guest/GuestTagsContainsKeywordsPredicate.java]: I remember AB4 used some disjoint set method to do something like this. Also possible to use stream here. But I guess this isn't a big issue for now.

File: src/main/java/seedu/address/model/guest/GuestTagsContainsKeywordsPredicate.java [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+
+        return false;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof GuestTagsContainsKeywordsPredicate // instanceof handles nulls
+                && keywords.equals(((GuestTagsContainsKeywordsPredicate) other).keywords)); // state check
+    }
```
Comment (teowz46) [inline src/main/java/seedu/address/model/guest/GuestTagsContainsKeywordsPredicate.java]: I think it's standard practice to override hashCode whenever you override equals.

File: src/main/java/seedu/address/model/guest/GuestPhoneExactPredicate.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+        return guest.getPhone().toString().equals(phoneNumber);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof GuestPhoneExactPredicate // instanceof handles nulls
+                && phoneNumber.equals(((GuestPhoneExactPredicate) other).phoneNumber)); // state check
+    }
+
```
Comment (teowz46) [inline src/main/java/seedu/address/model/guest/GuestPhoneExactPredicate.java]: I think need to override hashCode

File: src/main/java/seedu/address/model/guest/GuestEmailExactPredicate.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+        return guest.getEmail().toString().equals(email);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof GuestEmailExactPredicate // instanceof handles nulls
+                && email.equals(((GuestEmailExactPredicate) other).email)); // state check
+    }
+
```
Comment (teowz46) [inline src/main/java/seedu/address/model/guest/GuestEmailExactPredicate.java]: override hashCode

File: src/main/java/seedu/address/model/room/RoomBookingsDateRangePredicate.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+        return true;
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof RoomBookingsDateRangePredicate // instanceof handles nulls
+                && bookingPeriod.equals(((RoomBookingsDateRangePredicate) other).bookingPeriod)); // state check
+    }
+}
```
Comment (teowz46) [inline src/main/java/seedu/address/model/room/RoomBookingsDateRangePredicate.java]: hashcode

File: src/main/java/seedu/address/model/room/RoomCapacityExactPredicate.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+        return room.getCapacity().value == Integer.parseInt(capacity);
+    }
+
+    @Override
+    public boolean equals(Object other) {
+        return other == this // short circuit if same object
+                || (other instanceof RoomCapacityExactPredicate // instanceof handles nulls
+                && capacity.equals(((RoomCapacityExactPredicate) other).capacity)); // state check
+    }
+}
```
Comment (teowz46) [inline src/main/java/seedu/address/model/room/RoomCapacityExactPredicate.java]: i won't comment on anymore hashcodes lol, please check yourself. A methods you might find useful would be Objects.hash()

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java [hunk 10]
```diff
@@ -1,34 +1,185 @@
...
+import seedu.address.model.room.booking.BookingPeriod;
 
 public class FindCommandParserTest {
 
     private FindCommandParser parser = new FindCommandParser();
 
     @Test
     public void parse_emptyArg_throwsParseException() {
-        assertParseFailure(parser, "     ", String.format(MESSAGE_INVALID_COMMAND_FORMAT, FindCommand.MESSAGE_USAGE));
+        assertParseFailure(parser, "     ",
```
Comment (teowz46) [inline src/test/java/seedu/address/logic/parser/FindCommandParserTest.java]: I think there's a PREAMBLE_WHITESPACE static variable you can use here

**Pull request merged**